### PR TITLE
Header method

### PR DESCRIPTION
**Pull request by mohithapani**

### CODE REVIEW

File: here-oauth-client/src/main/java/com/here/account/http/HttpProvider.java [hunk 1]
```diff
@@ -98,6 +100,16 @@ void authorize(HttpRequest httpRequest, String method, String url,
...
          */
         InputStream getResponseBody() throws IOException;
+
+        /**
+         * Returns all the headers from the response
+         * @return returns an array of headers if the method implementation returns the headers or null if the method
+         *      is not implemented
+         */
+
+        default Header[] getHeaders() {
```
Comment (owenkellett) [inline here-oauth-client/src/main/java/com/here/account/http/HttpProvider.java]: You should provide a more generic type that does not pull in any Apache dependencies.  Generically headers would match up with a Map<String, List<String>>.  This is an option to use instead.

File: here-oauth-client/src/main/java/com/here/account/http/apache/ApacheHttpClientProvider.java [hunk 2]
```diff
@@ -161,9 +163,16 @@ public void addAuthorizationHeader(String value) {
     private static class ApacheHttpClientResponse implements HttpResponse {
         
         private final org.apache.http.HttpResponse apacheHttpResponse;
+        //private final Map<String, String> headers;
```
Comment (owenkellett) [inline here-oauth-client/src/main/java/com/here/account/http/apache/ApacheHttpClientProvider.java]: Do not include commented out code.

File: here-oauth-client/src/main/java/com/here/account/http/apache/ApacheHttpClientProvider.java [hunk 3]
```diff
@@ -161,9 +163,16 @@ public void addAuthorizationHeader(String value) {
...
         
         private ApacheHttpClientResponse(org.apache.http.HttpResponse apacheHttpResponse) {
             this.apacheHttpResponse = apacheHttpResponse;
+            //this.headers = new HashMap<>();
```
Comment (owenkellett) [inline here-oauth-client/src/main/java/com/here/account/http/apache/ApacheHttpClientProvider.java]: Do not include commented out code.

File: here-oauth-client/src/main/java/com/here/account/http/apache/ApacheHttpClientProvider.java [hunk 4]
```diff
@@ -161,9 +163,16 @@ public void addAuthorizationHeader(String value) {
...
+        }
+
+        private ApacheHttpClientResponse(org.apache.http.HttpResponse apacheHttpResponse, Map<String,String> headers) {
+            this.apacheHttpResponse = apacheHttpResponse;
+            //this.headers = headers;
```
Comment (owenkellett) [inline here-oauth-client/src/main/java/com/here/account/http/apache/ApacheHttpClientProvider.java]: Do not include commented out code.

**Comment by mohithapani** [comment]
Required changes made based on the comments.

File: here-oauth-client/src/main/java/com/here/account/http/HttpProvider.java [hunk 5]
```diff
@@ -98,6 +98,16 @@ void authorize(HttpRequest httpRequest, String method, String url,
...
         InputStream getResponseBody() throws IOException;
+
+        /**
+         * Returns all the headers from the response
+         * @return returns a Map of headers if the method implementation returns the headers or null if the method
+         *      is not implemented
+         */
+
+        default Map<String, List<String>> getHeaders() {
+            return null;
```
Comment (owenkellett) [inline here-oauth-client/src/main/java/com/here/account/http/HttpProvider.java]: I think it is more appropriate to throw UnsupportedOperationException here.  This is purely an optional method that adds convenience to the end user of the library.

File: here-oauth-client/src/main/java/com/here/account/http/apache/ApacheHttpClientProvider.java [hunk 6]
```diff
@@ -198,6 +201,24 @@ public long getContentLength() {
...
+        @Override
+        public Map<String, List<String>> getHeaders() {
+            Header[] headers = apacheHttpResponse.getAllHeaders();
+            Map<String, List<String>> ret = new HashMap<>();
+            for (Header header: headers) {
+                if(ret.containsKey(header.getName())) {
+                    ret.get(header.getName()).add(header.getValue());
+                }
+                else {
+                    ret.put(header.getName(), new ArrayList<>(Arrays.asList(header.getValue())));
```
Comment (owenkellett) [inline here-oauth-client/src/main/java/com/here/account/http/apache/ApacheHttpClientProvider.java]: No need to double instantiate.  Remove new ArrayList<> and just use Arrays.asList

File: here-oauth-client/src/main/java/com/here/account/http/java/JavaHttpProvider.java [hunk 7]
```diff
@@ -223,6 +235,7 @@ public HttpRequest getRequest(HttpRequestAuthorizer httpRequestAuthorizer, Strin
     
     protected long getContentLength(HttpURLConnection connection) {
         String contentLengthString = connection.getHeaderField(HttpConstants.CONTENT_LENGTH_HEADER);
+        connection.getHeaderFields();
```
Comment (owenkellett) [inline here-oauth-client/src/main/java/com/here/account/http/java/JavaHttpProvider.java]: What is this line for?

File: here-oauth-client/src/test/java/com/here/account/http/java/JavaHttpProviderTest.java [hunk 8]
```diff
@@ -333,5 +335,18 @@ public InputStream getInputStream() throws IOException {
...
+    public void test_JavaHttpClientResponse_Headers() throws HttpException, IOException {
+        String requestBodyJson = "{\"foo\":\"bar\"}";
+        String url = "http://google.com";
+        HttpProvider httpProvider = JavaHttpProvider.builder().build();
+        HttpRequest httpRequest = httpProvider.getRequest(httpRequestAuthorizer, "PUT", url, requestBodyJson);
+        HttpProvider.HttpResponse response = httpProvider.execute(httpRequest);
+        assertEquals(HttpURLConnection.HTTP_BAD_METHOD, response.getStatusCode());
+        assertNotNull("response body is null", response.getResponseBody());
+        assertTrue("response content length is 0", 0<response.getContentLength());
+        assertTrue("Header method should not return null",response.getHeaders() != null);
```
Comment (owenkellett) [inline here-oauth-client/src/test/java/com/here/account/http/java/JavaHttpProviderTest.java]: Maybe check for a specific header that you expect to be there?

**Pull request merged**